### PR TITLE
Restructure build (#474)

### PR DESCRIPTION
**Pull request by aepfli**
Unifying builds into one pipeline with configurable options.
Removing all other now unnecessary builds.
Adding experimental builds.

Closes: #474

Proposed commit message:

```
Restructure build into a mono build (#474 / #495)

Instead of multiple smaller builds, we introduces one mono build. This
allows us to provide a short and useful feedback cycle for
contributors, but ensures safety regarding merging for maintainers.
As for maintainers it does not matter how long our build runs before
a merge, and in the end it does not matter if it takes hours.

Furthermore our release build is now part of the pipeline and
we do not need to wait for another build for releasing.

Closes: #474 
PR: #495
```

---
**PR checklist**

The following checklist shall help the PR's author, the reviewers and maintainers to ensure the quality of this project.
It is based on our contributors guidelines, especially the ["writing code" section](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#writing-code).
It shall help to check for completion of the listed points.
If a point does not apply to the given PR's changes, the corresponding entry can be simply marked as done. 

Documentation (general)
* [ ] There is documentation (Javadoc and site documentation; added or updated)
* [ ] There is implementation information to describe _why_ a non-obvious source code / solution got implemented
* [ ] Site documentation has its own `.adoc` file in the `docs` folder, e.g. `docs/report-entries.adoc`
* [ ] Only one sentence per line (especially in `.adoc` files)
* [ ] Javadoc uses formal style, while sites documentation may use informal style

Documentation (new extension)
* [ ] The `docs/docs-nav.yml` navigation has an entry for the new extension
* [ ] The `package-info.java` contains information about the new extension

Code
* [ ] Code adheres to code style, naming conventions etc.
* [ ] Successful tests cover all changes
* [ ] There are checks which validate correct / false usage / configuration of a functionality and there are tests to verify those checks
* [ ] Tests use [AssertJ](https://assertj.github.io/doc/) or our own [PioneerAssert](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#assertions) (which are based on AssertJ)

Contributing
* [ ] A prepared commit message exists
* [ ] The list of contributions inside `README.md` mentions the new contribution (real name optional) 

---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

File: .github/workflows/build.yml
Comment (aepfli) [inline .github/workflows/build.yml:60]: maybe we should just call that `Basic on ${{ matrix.os }}`

**Comment by beatngu13** [comment]
So far, I haven't had time to take a closer look – sorry! ✌️

What do you think of a _short_ description of that build/PR process in `CONTRIBUTING.md`? (New) contributors might be a bit confused when they see a failing check, even if the job/step names are already quite descriptive.

File: .github/workflows/build.yml
Comment (Bukama) [reply .github/workflows/build.yml:60]: But then you don't see that Java version "Basic" is

**Comment by aepfli** [comment]
it might be even worth to not break the build but set up a dependent build step and add a stage `ready to merge` at the end, like we had for before for `buildOk`. we can than setup the project to rely on that build before merging. which might be even better than a failing build. This seems to be less intrusive and easier to understand.

**Comment by aepfli** [comment]
Okay i tried with "needed workflows" but as soon as one is skipped on purpose, it will still not block a merge.

I will adapt the CONTRIBUTORS.MD as i see no better way arround this for for now

**Comment by nipafx** [comment]
I really like the entire setup - it looks amazing in GitHub's workflow visualization. Great job, @aepfli!

Only thing I want to find a better solution for is how to prevent premature merges (i.e. before running the full build). At the moment, the full build is triggered with by adding a label _merge-ready_ and because that can be somewhat easily forgotten, the step _Maintainer Action is Missing before Merge_ checks that and fails if it's absent. It works, but I don't like the non-technical implications. Contributors will get mails for failing builds and the PR overview will be mostly ❌. That's a lot of negative feedback even though nobody made a mistake.

Are there other options we can explore? What about defaulting to leaving the PR in draft mode (which runs the small build suite) until all is said and done and then switch to "proper" PR (which runs all the builds)? If we find no other option, we could also simply remove that step and hope that we don't forget the label too often (adding a checkbox to the default PR message would help).

**Comment by aepfli** [comment]
based on @filiphr input (https://github.com/marketplace/actions/github-checks) i am experimenting with this action.

We do have multiple options here, like:
- set it to in_progress:
    Will never finish might be also bad feedback
- set to action_needed:
    Will cause an email still the question if we can work with that see https://github.com/aepfli/junit-pioneer/pull/11
- set it to neutral:
    will not prevent a merge, it will be just grey :(

**Comment by aepfli** [comment]
okay for `action_required` i did not receive an email. which is good - and at least resolves the feedback for the contributors. but it will still  show an red `x` beside the PR

**Comment by Bukama** [comment]
TBH I don't see in what state this is and what's your final thoughts after your last concerns :/

**Comment by aepfli** [comment]
My last concerns summarized:

- Action needed still shows a red X on the merge request in the overview - but it does not send an email
- Label based approach works, but as soon as the label was added once, our condition is always evaluating to true. I would leave this for now, and define this as a future improvement.
- The build works now as intended, and i would suggest to give this a go. As always testing CI stuff is hard, so i expect issues, but i did not find any in my fork for now.

File: .github/workflows/build.yml
Comment (beatngu13) [inline .github/workflows/build.yml:204]: Hmmm, why this additional Java setup with v16 and AdoptOpenJDK?

File: .github/workflows/build.yml
Comment (beatngu13) [inline .github/workflows/build.yml:223]:
```suggestion
          arguments: --refresh-dependencies -PexperimentalJavaVersion=${{ env.EXPERIMENTAL_JAVA }} -PmodularBuild=${{ matrix.modular }} -PjunitVersion=${{ matrix.junit-version }} -Porg.gradle.java.installations.auto-download=false --stacktrace --scan clean build
```

File: .github/workflows/build.yml
Comment (beatngu13) [inline .github/workflows/build.yml:237]: A comment would be helpful – just like in the case of Sonar – explaining why unshallow is needed (I also can't remember right now). Also, above we use an integer, here string.

File: .github/workflows/build.yml
Comment (beatngu13) [inline .github/workflows/build.yml:259]: This job key uses camel case, but there's also e.g. `full-featured`.

File: .github/workflows/build.yml
Comment (beatngu13) [inline .github/workflows/build.yml:262]: All timeouts seem to be 15 min, maybe define a envvar `DEFAULT_JOB_TIMEOUT`?

File: build.gradle.kts
Comment (beatngu13) [inline build.gradle.kts:28]:
```suggestion
val experimentalBuild: Boolean = experimentalJavaVersion?.isNotEmpty() ?: false
```

File: build.gradle.kts
Comment (beatngu13) [inline build.gradle.kts:27]: Elsewhere, we don't use semicolons.

File: build.gradle.kts
Comment (beatngu13) [inline build.gradle.kts:31]:
```suggestion
	if (experimentalBuild) {
```

File: build.gradle.kts
Comment (beatngu13) [inline build.gradle.kts:36]:
```suggestion
		if (modularBuild.toBoolean()) {
```

(We need a Spotless config for Kotlin, just like we used to have with Groovy: #484)

File: build.gradle.kts
Comment (beatngu13) [inline build.gradle.kts:207]:
```suggestion
    	    	configure<JacocoTaskExtension> {
```

**Review by beatngu13 — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml
Comment (aepfli) [reply .github/workflows/build.yml:8]: true, i should have looked better

File: .github/workflows/build.yml
Comment (aepfli) [reply .github/workflows/build.yml:27]: would be, but this is some kind of safety feature, you have to double set something, so this happening by accident is reduced

File: .github/workflows/build.yml
Comment (aepfli) [reply .github/workflows/build.yml:57]:
```suggestion
```

File: .github/workflows/build.yml
Comment (aepfli) [reply .github/workflows/build.yml:204]: gradle is not working with ea builds - it will fail, so we start gradle with a suported java version, but execute the build on the experimental version

File: .github/workflows/build.yml
Comment (aepfli) [reply .github/workflows/build.yml:237]:
```suggestion
          # Sonar needs the whole git history for issue assignment
          fetch-depth: '0'
```

File: .github/workflows/build.yml
Comment (aepfli) [reply .github/workflows/build.yml:259]:
```suggestion
  update-website:
```

File: .github/workflows/build.yml
Comment (aepfli) [reply .github/workflows/build.yml:262]: that is a good idea - i will add this as soon as i have time

File: .github/workflows/build.yml
Comment (aepfli) [reply .github/workflows/build.yml:133]:
```suggestion
          action_url: "https://github.com/junit-pioneer/junit-pioneer/blob/main/CONTRIBUTING.md"
```

File: .github/workflows/build.yml
Comment (beatngu13) [reply .github/workflows/build.yml:27]: Let's go one step further and implement a [two-man rule](https://en.wikipedia.org/wiki/Two-man_rule). 😉

Personally, I wouldn't need this, but it's fine.

File: .github/workflows/build.yml
Comment (beatngu13) [reply .github/workflows/build.yml:203]: Would a comment make sense then?

File: .github/workflows/build.yml
Comment (beatngu13) [reply .github/workflows/build.yml]: I understand why we need it for the `sonarcloud` job, but why for the `release` job?

Comment (aepfli) [reply .github/workflows/build.yml]: for changelog generation

**Review by Bukama — APPROVED** [review]
Everything green :D
So technical it works, but @beatngu13  and @aepfli I ask you to finish the open points.

**Review by Michael1993 — APPROVED** [review]
Everything is green. 😄

File: build.gradle.kts
Comment (aepfli) [reply build.gradle.kts:27]:
```suggestion
val modularBuild : String by project
val experimentalJavaVersion : String? by project
```

**Comment by nipafx** [comment]
I really liked the idea to decide between basic and full build based on the PR's draft state, but that requires all PRs to be draft by default and most of their life time. This is counter to how they are used elsewhere on GitHub (and even in this project so far) and will probably lead to most new contributors opening "regular" PRs, which means the get the full build. I'm not sure that's such a big issue: most contributors aren't new and we can easily turn the PR into a draft and let them know about out practice; then again, anything non-standard will always lead to friction. For now we'll not use this approach.

File: .github/workflows/build.yml
Comment (nipafx) [reply .github/workflows/build.yml:8]: Shipkit doesn't push commits, just tags. If that triggers a _push_ event, we need to make sure we don't build on new _v*_ tags. Or really any tags. Then again, if a new commit gets pushed, no matter what tags it comes with, the build needs to be executed.

File: .github/workflows/build.yml
Comment (nipafx) [reply .github/workflows/build.yml:27]: The reason for having that boolean flag is that I found it better to make explicit whether one wants to release or not.

**Comment by aepfli** [comment]
Here is a little overview of different actions for different builds on GitHub pull request overview - https://github.com/aepfli/junit-pioneer/pulls

Here are my thoughts on those:

- action_required - here is the overview of the action requiered https://github.com/aepfli/junit-pioneer/pull/12/checks?check_run_id=3001483599 - it is a red cross but no email

- with queued, it will always stay yellow, if that is a better solution i am not sure :)

File: .github/workflows/build.yml
Comment (nipafx) [reply .github/workflows/build.yml:130]: What @beatngu13 said!

File: build.gradle.kts
Comment (nipafx) [reply build.gradle.kts:27]: Fixed.

**Comment by nipafx** [comment]
It's only us maintainers who can merge a PR and maybe we can just remember to apply the label before actually merging. So let's go with the simplest solution for now: No mechanism preventing accidental merges that didn't get a full build. If it turns out that we make too many mistakes, we can still add something later.

**Comment by nipafx** [comment]
As far as I can tall, only [this comment](https://github.com/junit-pioneer/junit-pioneer/pull/495#discussion_r664754199) needs to be resolved before this can be merged. Can you take a look, @aepfli? 😁

File: .github/workflows/build.yml
Comment (nipafx) [reply .github/workflows/build.yml:59]: Agreed, lets leave the Java version in there.

File: .github/workflows/build.yml
Comment (nipafx) [reply .github/workflows/build.yml]: That section got removed.

File: .github/workflows/build.yml
Comment (nipafx) [reply .github/workflows/build.yml:185]: I added a comment.

File: .github/workflows/build.yml
Comment (nipafx) [reply .github/workflows/build.yml:244]: I tried, it lead to the build not starting...

**Comment by nipafx** [comment]
@beatngu13 Can you take a look whether the changes you requested are done?

File: .github/workflows/build.yml
Comment (beatngu13) [reply .github/workflows/build.yml:244]: Thx for trying it out, didn't know/expect that.

As pointed out by @aepfli during the stream, I believe we can use [`defaults`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#defaults), "[a] map of default settings that will apply to all jobs in the workflow".

```yaml
defaults:
  timeout-in-minutes: 15
```

This way, we should be able to remove the timeouts from all jobs.

**Review by beatngu13 — APPROVED** [review]

**Comment by beatngu13** [comment]
One optional improvement we could try out:

https://github.com/junit-pioneer/junit-pioneer/pull/495#discussion_r667385417

But anyway, great job – approved! 🚀

File: .github/workflows/build.yml
Comment (nipafx) [reply .github/workflows/build.yml:244]: See 7304aae.

**Review by nipafx — APPROVED** [review]
LGTM

**Pull request merged**
